toolbox/metrics/metrics.go

```diff
@@ -0,0 +1,82 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics
+
+import (
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/net/context"
+)
+
+// Suite of metrics to collect from.
+type Suite map[string]Metric
+
+// NewPublisher creates a publisher from a suite and update interval and timeout
+func NewPublisher(s Suite, updateInterval, updateTimeout time.Duration) *Publisher {
+	return &Publisher{
+		suite:          s,
+		updateInterval: updateInterval,
+		updateTimeout:  updateTimeout,
+	}
+}
+
+// Publisher manages publication of multiple metrics
+type Publisher struct {
+	suite                         Suite
+	updateInterval, updateTimeout time.Duration
+}
+
+// Metric is used to collect data
+type Metric interface {
+	Update(ctx context.Context) error
+	GetCollector() prometheus.Collector
+}
+
+// RegisterMetrics should be called in init
+func (p *Publisher) RegisterMetrics() {
+	for _, m := range p.suite {
```

Comment:
Call `RegisterMetrics` inside `metrics.NewPublisher()` as part of processing `suite`. AKAICT, all usage of `Publisher` in this PR requires `RegisterMetrics()`.